src/SalesShareDistributor.sol

```diff
@@ -0,0 +1,180 @@
+
+import { IERC1155Supply } from "./IERC1155Supply.sol";
+
+import { FractionalizedToken, Metadata } from "./FractionalizedToken.sol";
+
+import { SchmackoSwap, ListingState } from "./SchmackoSwap.sol";
+import { IPermissioner, TermsAcceptedPermissioner } from "./Permissioner.sol";
+
+struct Sales {
+    uint256 fulfilledListingId;
+    IERC20 paymentToken;
+    uint256 paidPrice;
+    IPermissioner permissioner;
+}
+
+error ListingNotFulfilled();
+error ListingMismatch();
+error InsufficientBalance();
+error UncappedToken();
+error OnlyIssuer();
+
+error NotClaimingYet();
+
+contract SalesShareDistributor is UUPSUpgradeable, OwnableUpgradeable, ReentrancyGuardUpgradeable {
+    using SafeERC20 for IERC20;
+
+    SchmackoSwap private schmackoSwap;
+
+    mapping(address => Sales) public sales;
+
+    event SalesActivated(address indexed fractionToken, address paymentToken, uint256 paidPrice);
+    event SharesClaimed(address indexed fractionToken, address indexed claimer, uint256 amount);
+
+    function initialize(SchmackoSwap _schmackoSwap) public {
+        schmackoSwap = _schmackoSwap;
+    }
+
+    /**
+     * @notice returns the `amount` of `paymentToken` that `tokenHolder` can claim by burning their fractions
+     *
+     * @param tokenContract address
+     * @param holder address
+     */
+    function claimableTokens(FractionalizedToken tokenContract, address holder) public view returns (IERC20 paymentToken, uint256 amount) {
+        Sales memory _sales = sales[address(tokenContract)];
+
+        if (address(_sales.paymentToken) == address(0)) {
+            revert NotClaimingYet();
+        }
+
+        uint256 balance = tokenContract.balanceOf(holder);
+
+        return (_sales.paymentToken, (balance * _sales.paidPrice) / tokenContract.totalIssued());
+    }
+
+    /**
+     * @notice call during claiming phase to burn all fractions and receive the pro rata sales share
+     * @param permissions bytes data that can be read and verified by the configured permissioner
+     *        at the moment this simply is a valid signature over a `specificTermsV1` message
+     */
+    function claim(FractionalizedToken tokenContract, bytes memory permissions) public {
+        uint256 balance = tokenContract.balanceOf(_msgSender());
+        if (balance == 0) {
+            revert InsufficientBalance();
+        }
+        Sales memory _sales = sales[address(tokenContract)];
+
+        (IERC20 paymentToken, uint256 erc20shares) = claimableTokens(tokenContract, _msgSender());
+
+        _sales.permissioner.accept(tokenContract, _msgSender(), permissions);
+
+        if (erc20shares == 0) {
+            //todo: this is very hard to simulate because the condition above will already yield 0
+            revert InsufficientBalance();
+        }
+        emit SharesClaimed(address(tokenContract), _msgSender(), balance);
+        tokenContract.burnFrom(_msgSender(), balance);
+        paymentToken.safeTransfer(_msgSender(), erc20shares);
```

Comment:
What's the reason here for checking the balance early instead of checking it after getting the `claimableTokens` which will then return the balance of 0?

Reply:
might be a a relic of the old code. It likely works with a plain claimable amt check.

Reply:
you need the balance to burn and emit again.